crates/ruff/src/rules/flake8_comprehensions/rules/unnecessary_comprehension.rs

```diff
@@ -43,19 +44,36 @@ pub fn unnecessary_comprehension(
     if !(generator.ifs.is_empty() && generator.is_async == 0) {
         return;
     }
-    let Some(elt_id) = helpers::function_name(elt) else {
-        return;
-    };
+    let elt_ids: Vec<Option<&str>> = elts.iter().map(|&e| helpers::function_name(e)).collect();
 
-    let Some(target_id) = helpers::function_name(&generator.target) else {
-        return;
+    let target_ids: Vec<Option<&str>> = match &generator.target.node {
```

Comment:
Can we try to split this into two separate cases: `List` and `Set` vs. `Dict`? Since for `List` and `Set`, I believe we _always_ want a `Name` here, whereas for `Dict`, we _always_ want a length-two tuple. (Looking at the [flake8-comprehensions implementation](https://github.com/adamchainz/flake8-comprehensions/pull/490/files#diff-2fafc059a4a4fef0240cd305df8b843c6937ce529a7905567860392cd1cc76d4R311) for reference.) Right now, those expectations are all implicit.